src/WebJobs.Script.WebHost/FileMonitoringService.cs

```diff
@@ -126,6 +117,32 @@ private void InitializeFileWatchers()
                     .Subscribe((msg) => ScheduleRestartAsync(false)
                     .ContinueWith(t => _logger.LogCritical(t.Exception.Message),
                         TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously)));
+
+            // Delay starting up for logging and debug file watchers to avoid long start up times
+            Utility.ExecuteAfterDelay(InitializeNonBlockingFileWatchers, TimeSpan.FromSeconds(_delayedWatchersDelaySeconds));
+        }
+
+        /// <summary>
+        /// Initializes the file and directory monitoring that does not need to happen as part of a Host startup
+        /// These watchers can be started after a delay to avoid startup performance issue
+        /// </summary>
+        private void InitializeNonBlockingFileWatchers()
+        {
+            lock (_stopWatchersLock)
```

Comment:
Perform a check here before attempting to acquire the lock as well. Avoid locking/waiting if not needed.